File: crates/convertor/src/url/conv_url.rs
use crate::common::encrypt::Encryptor;
use crate::error::ConvUrlError;
use crate::url::conv_query::ConvQuery;
use serde::{Deserialize, Serialize};
use std::fmt::{Display, Formatter};
use std::str::FromStr;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConvUrl {
    r#type: UrlType,
    server: url::Url,
    query: Option<ConvQuery>,
}

impl ConvUrl {
    pub fn new(r#type: UrlType, server: url::Url, query: Option<ConvQuery>) -> Self {
        Self { r#type, server, query }
    }

    pub fn empty() -> Self {
        Self {
            r#type: UrlType::Original,
            server: url::Url::parse("http://example.com").unwrap(),
            query: None,
        }
    }

    pub fn original(url: url::Url) -> Self {
        Self {
            r#type: UrlType::Original,
            server: url,
            query: None,
        }
    }

    pub fn take_query(mut self) -> Result<ConvQuery, ConvUrlError> {
        self.query.take().ok_or(ConvUrlError::MissingConvQuery)
    }

    pub fn encrypt(self, encryptor: &Encryptor) -> Result<Self, ConvUrlError> {
        let Self { r#type, server, query } = self;
        let query = query
            .map(|q| q.encrypt(encryptor))
            .transpose()
            .map_err(ConvUrlError::EncryptQuery)?;
        Ok(Self { r#type, server, query })
    }

    pub fn decrypt(self, encryptor: &Encryptor) -> Result<Self, ConvUrlError> {
        let Self { r#type, server, query } = self;
        let query = query
            .map(|q| q.decrypt(encryptor))
            .transpose()
            .map_err(ConvUrlError::EncryptQuery)?;
        Ok(Self { r#type, server, query })
    }
}

impl TryFrom<&ConvUrl> for url::Url {
    type Error = ConvUrlError;

    fn try_from(value: &ConvUrl) -> Result<Self, Self::Error> {
        let mut url = value.server.clone();
        if matches!(value.r#type, UrlType::Original) {
            return Ok(url);
        }
        url.set_path(value.r#type.path());
        let query: Option<String> = value
            .query
            .as_ref()
            .map(|q| q.try_into())
            .transpose()
            .map_err(ConvUrlError::ConvQuery)?;
        url.set_query(query.as_deref());
        Ok(url)
    }
}

impl TryFrom<ConvUrl> for url::Url {
    type Error = ConvUrlError;

    fn try_from(value: ConvUrl) -> Result<Self, Self::Error> {
        url::Url::try_from(&value)
    }
}

impl TryFrom<&url::Url> for ConvUrl {
    type Error = ConvUrlError;

    fn try_from(value: &url::Url) -> Result<Self, Self::Error> {
        value.as_str().parse()
    }
}

impl TryFrom<url::Url> for ConvUrl {
    type Error = ConvUrlError;

    fn try_from(value: url::Url) -> Result<Self, Self::Error> {
        ConvUrl::try_from(&value)
    }
}

impl FromStr for ConvUrl {
    type Err = ConvUrlError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let mut server = s.parse::<url::Url>().map_err(|e| ConvUrlError::InvalidUrl(e.to_string(), e))?;
        let r#type = UrlType::from_path(server.path());
        let query = if !matches!(r#type, UrlType::Original) {
            let query = server
                .query()
                .map(|query| query.parse::<ConvQuery>().map_err(ConvUrlError::ParseQuery))
                .transpose()?;
            server.set_path("");
            server.set_query(None);
            query
        } else {
            None
        };
        Ok(Self { r#type, server, query })
    }
}

impl Display for ConvUrl {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let mut url = self.server.clone();
        if !matches!(self.r#type, UrlType::Original) {
            url.set_path(self.r#type.path());
            if let Some(query) = &self.query {
                url.set_query(Some(query.to_string().as_str()));
            }
        }
        write!(f, "{}", url)
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[derive(Serialize, Deserialize)]
pub enum UrlType {
    Original,
    Raw,
    Profile,
    ProxyProvider,
    RuleProvider,
}

impl UrlType {
    pub fn label(&self) -> String {
        match self {
            UrlType::Original => "订阅商原始订阅配置".to_string(),
            UrlType::Raw => "转换前订阅配置".to_string(),
            UrlType::Profile => "转换后订阅配置".to_string(),
            UrlType::ProxyProvider => "代理提供者".to_string(),
            UrlType::RuleProvider => "规则提供者".to_string(),
        }
    }

    pub fn path(&self) -> &'static str {
        match self {
            UrlType::Original => "/api/original",
            UrlType::Raw => "/api/raw",
            UrlType::Profile => "/api/profile",
            UrlType::ProxyProvider => "/api/proxy-provider",
            UrlType::RuleProvider => "/api/rule-provider",
        }
    }

    pub fn from_path(path: &str) -> Self {
        match path {
            "/api/raw" => UrlType::Raw,
            "/api/profile" => UrlType::Profile,
            "/api/proxy-provider" => UrlType::ProxyProvider,
            "/api/rule-provider" => UrlType::RuleProvider,
            _ => UrlType::Original,
        }
    }
}

impl UrlType {
    pub fn variants() -> &'static [Self] {
        &[
            UrlType::Original,
            UrlType::Raw,
            UrlType::Profile,
            UrlType::ProxyProvider,
            UrlType::RuleProvider,
        ]
    }
}

impl Display for UrlType {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            UrlType::Original => write!(f, "raw"),
            UrlType::Raw => write!(f, "raw_profile"),
            UrlType::Profile => write!(f, "profile"),
            UrlType::ProxyProvider => write!(f, "proxy_provider"),
            UrlType::RuleProvider => write!(f, "rule_provider"),
        }
    }
}

File: dashboard-v2/src/app/common/model/core/conv-url.ts
import * as z from "zod";
import Cloneable from "../../base/cloneable";
import Equatable from "../../base/equals";
import Serializable from "../../base/serializable";
import ConvQuery, { ConvQuerySchema } from "./conv-query";

export enum UrlTypeVariant {
    Original = "Original",
    Raw = "Raw",
    Profile = "Profile",
    ProxyProvider = "ProxyProvider",
    RuleProvider = "RuleProvider",
}

const UrlTypeVariantSchema = z.enum(UrlTypeVariant, {
    error: (issue) => `Unknown UrlType: ${String(issue.input)}`,
});

export const ConvUrlSchema = z.object({
    type: UrlTypeVariantSchema,
    server: z.string(),
    query: ConvQuerySchema.nullable().optional(),
});

export type ConvUrlJson = z.infer<typeof ConvUrlSchema>;

export class ConvUrl implements Cloneable<ConvUrl>, Equatable<ConvUrl>, Serializable {
    private constructor(
        public readonly type: UrlType,
        public readonly server: string,
        public readonly query: ConvQuery | null,
    ) {
    }

    // --- Static constructors ---

    public static new(type: UrlType, server: string, query: ConvQuery | null): ConvUrl {
        return new ConvUrl(type, server, query);
    }

    public static empty(): ConvUrl {
        return new ConvUrl(UrlType.Original, "http://example.com", null);
    }

    public static original(server: string): ConvUrl {
        return new ConvUrl(UrlType.Original, server, null);
    }

    /** FromStr: parse a full URL string into ConvUrl */
    public static fromStr(s: string): ConvUrl {
        const url = new URL(s);
        const type = UrlType.fromPath(url.pathname);
        let query: ConvQuery | null = null;
        if (type !== UrlType.Original && url.search) {
            query = ConvQuery.fromSearch(url.search);
        }
        const server = `${url.protocol}//${url.host}`;
        return new ConvUrl(type, server, query);
    }

    // --- Instance methods ---

    /** path_and_query */
    public pathAndQuery(): string {
        const path = this.type.path();
        if (this.query === null) {
            return path;
        }
        return `${path}?${this.query.toString()}`;
    }

    /** takeQuery: consumes the query (returns it, leaving null conceptually) */
    public takeQuery(): ConvQuery {
        if (this.query === null) {
            throw new Error("MissingConvQuery");
        }
        return this.query;
    }

    /** TryFrom<ConvUrl> for URL */
    public toUrl(): URL {
        const url = new URL(this.server);
        if (this.type === UrlType.Original) {
            return url;
        }
        url.pathname = this.type.path();
        if (this.query !== null) {
            url.search = this.query.toString();
        }
        return url;
    }

    /** Display */
    public toString(): string {
        return this.toUrl().toString();
    }

    public desc(): string {
        return this.type.label;
    }

    // --- Cloneable / Equatable / Serializable ---

    public clone(): ConvUrl {
        return new ConvUrl(this.type.clone(), this.server, this.query);
    }

    public equals(other?: ConvUrl): boolean {
        if (!other) {
            return false;
        }
        return this.type.equals(other.type) && this.server === other.server;
    }

    public serialize(): any {
        return {
            type: this.type.serialize(),
            server: this.server,
            query: this.query ?? null,
        };
    }

    public static parse(json: unknown): ConvUrlJson {
        return ConvUrlSchema.parse(json);
    }

    public static deserialize(conv_url: ConvUrlJson): ConvUrl {
        return new ConvUrl(
            UrlType.deserialize(conv_url.type),
            conv_url.server,
            !!conv_url.query ? ConvQuery.deserialize(conv_url.query) : null,
        );
    }
}

export class UrlType implements Cloneable<UrlType>, Equatable<UrlType>, Serializable {
    public static readonly PREFIX = "/subscription";

    public static readonly Original = new UrlType(UrlTypeVariant.Original, "/original", "订阅商原始订阅配置", "original");
    public static readonly Raw = new UrlType(UrlTypeVariant.Raw, "/raw", "转换前订阅配置", "raw_profile");
    public static readonly Profile = new UrlType(UrlTypeVariant.Profile, "/profile", "转换后订阅配置", "profile");
    public static readonly ProxyProvider = new UrlType(UrlTypeVariant.ProxyProvider, "/proxy-provider", "代理提供者", "proxy_provider");
    public static readonly RuleProvider = new UrlType(UrlTypeVariant.RuleProvider, "/rule-provider", "规则提供者", "rule_provider");

    private constructor(
        public readonly variant: UrlTypeVariant,
        public readonly subPath: string,
        public readonly label: string,
        private readonly displayName: string,
    ) {
    }

    public static prefix(): string {
        return UrlType.PREFIX;
    }

    public path(): string {
        return UrlType.PREFIX + this.subPath;
    }

    public static variants(): UrlType[] {
        return [
            UrlType.Original,
            UrlType.Raw,
            UrlType.Profile,
            UrlType.ProxyProvider,
            UrlType.RuleProvider,
        ];
    }

    public static fromPath(path: string): UrlType {
        return UrlType.variants().find(v => v.path() === path) ?? UrlType.Original;
    }

    public toString(): string {
        return this.displayName;
    }

    // --- Cloneable / Equatable / Serializable ---

    public clone(): UrlType {
        return this; // 不可变单例，直接返回自身
    }

    public equals(other?: UrlType): boolean {
        if (!other) {
            return false;
        }
        return this.variant === other.variant;
    }

    public serialize(): any {
        return this.variant;
    }

    public static deserialize(value: UrlTypeVariant): UrlType {
        const type = UrlType.variants().find(v => v.variant === value);
        if (type === undefined) {
            throw new Error(`Unknown UrlType: ${String(value)}`);
        }
        return type;
    }
}

